keps/sig-storage/0037-20190129-PVCTaint.md

```diff
@@ -0,0 +1,123 @@
+  - n/a
+---
+
+# Title
+
+PVC/PV Taint and Toleration
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+The current Taint and Toleration feature allows users to mark a node (taint the node) so that no pods can be scheduled to it, unless a pod explicitly tolerates the taint. This proposal is trying to extend this concept on a volume. User or admin (or controller) can mark a volume (taint the PVC or PV) so that no pods could start to use them unless a pod can tolerate the taint. This feature is very useful for some storage operations such as volume snapsoht/clone and data population.  
+
+## Motivation
+
+Currently in our volume management design, k8s only supports provisioning an empty volume and a volume can be used by a pod immediately after it is created. With the new features on storage and data management becoming available such as snapshots, clone, and data population, data might be restored or populated to the volumes. We need a way to indicate when the data is available on the volume so that pod can start using it. 
+
+Previously, we proposed to use Readiness Gate for this problem, however, this approach lacks the support the scenarios that some special Pods (e.g., Pods for populating the data) can be allowed to use the volume even when the data is not ready.
+
+Taint and toleration feature currently used for marking node and reacting based
+on the taint effect seems also provides a nice way to solve this volume issue.
+
+### Goals
+
+This proposal is trying to investigate the taint and toleration feature applied
+to PVC/PV. It will identify the use cases that this feature could be applied to
+and describe the workflow of how this feature can be used.
+
+### Non-Goals
+
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects: 
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+These effects are used for tainting the nodes so that scheduler or controller
+will take actions on Pods based on the effect. Here we propose a new effect, TaintEffectNoAttachOrMount. For a Pod that cannot tolerate this taint, attach_detach controller and volume manager should skip this volume and not trigger attach or mount operations.
```

Comment:
Why re-use the same definitions?  What does this new taint mean when applied to a node?  What do the old taints mean when applied to a PV?

This feels a different enum.

Likewise, I am not sure it should be in the same field on the pod.

Reply:
I think choose to use different definitions has both procs and cons. The current taint and tolerations APIs have very general names which do not explicitly indicate it is only used for node. If we propose a different sets of taints and tolerations for volumes and pods, we basically have two identical data struct with different names. If in the future, if there are other APIs that might use taint and tolerations feature, we have to add more such data structs. 

I am thinking although the PVC/PV taints are used for different objects, but the way it is used is basically the same as nodes. User can taint something which trigger controller to have some actions based on the effect. The TaintEffect name kind of indicates that this taint should be used on what types of objects.

Reply:
I agree there are tradeoffs.  My personal feeling is that this is the less desirable path.  I would want to discuss with other API reviewers, and to get feedback from the other side of this (scheduling).

@liggitt @smarterclayton @bsalamat

Reply:
I agree with @thockin. The existing Taint and Toleration is more accurately called "Node" taint and toleration, not "generic" taint and toleration that could be applied to other objects. I don't object introducing "PV/PVC taint and toleration", but if we go that path, we should know that this new taint and toleration will be a new API which shares only part of the name and the concept, but not the existing behavior. In other words, there is little point in saying that the existing taint and toleration supports X and Y taint effect and we are adding Z. Whatever taint effects is supported by "node taint" won't be applicable to "volume taint". You are creating a new API and the new API should support its own type of taints that makes sense in the context of volumes and volume binding.

Reply:
Thanks for the feedback. Considering the use cases @msau42 and others mentioned, besides NoAttachOrMount effect for PVC taint, we could have other effects such as EvictPod when PVC is tainted. To avoid the confusion between PVC taint effect and Node taint effect, I think it is better to have a separate set of taint and toleration for PVC/PV.

Another option is to add a Kind field into the Taint and tolerations to indicate this taint should be applied to what types of objects, node, or pvc or others. It has to be optional, and default type should be node. But this way, it might still cause confusion on TaintEffect, it might not be very clear for users which TaintEffect should apply to what kind of objects.